Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```diff
@@ -68,14 +73,82 @@
+        completionHandler(true);
+        return;
+    }
+
     parseMatchPatterns(origins, matchPatterns);
 
     if (hasPermissions(permissions, matchPatterns)) {
         completionHandler(true);
         return;
     }
 
-    // FIXME: <https://webkit.org/b/250135> Call delegate method to prompt user for access.
+    // There shouldn't be any unsupported permissions, since they got reported as an error before this.
+    ASSERT(permissions.isSubset(extension().supportedPermissions()));
+
+    bool requestedAllHostsPattern = false;
+    for (auto& matchPattern : matchPatterns) {
+        if (matchPattern->matchesAllHosts()) {
+            requestedAllHostsPattern = true;
+            break;
+        }
+    }
+
+    if (requestedAllHostsPattern && !m_requestedOptionalAccessToAllHosts)
+        m_requestedOptionalAccessToAllHosts = YES;
+
+    __block MatchPatternSet grantedPatterns;
+    __block PermissionsSet grantedPermissions = permissions;
+    auto delegate = extensionController()->wrapper().delegate;
+
+    auto originsCompletionHandler = ^(NSSet<_WKWebExtensionMatchPattern *> *allowedOrigins) {
+        for (_WKWebExtensionMatchPattern *pattern in allowedOrigins) {
+            // Allowed origin doesn't match the origins requested.
+            if (!origins.contains(pattern.string)) {
+                completionHandler(false);
+                return;
+            }
+
+            grantedPatterns.add(*WebExtensionMatchPattern::getOrCreate(pattern.string));
+        }
+
+        grantPermissionMatchPatterns(WTFMove(grantedPatterns));
+        grantPermissions(WTFMove(grantedPermissions));
+        completionHandler(true);
+    };
+
+    auto permissionCompletionHandler = ^(NSSet<_WKWebExtensionPermission> *allowedPermissions) {
+        for (_WKWebExtensionPermission permission in allowedPermissions) {
+            // Allowed permission doesn't match the permissions requested.
+            if (!grantedPermissions.contains(permission)) {
+                completionHandler(false);
+                return;
+            }
+        }
+
+        if ([delegate respondsToSelector:@selector(webExtensionController:promptForPermissionMatchPatterns:inTab:forExtensionContext:completionHandler:)]) {
+            NSMutableSet *matchPatternsToRequest = [NSMutableSet set];
+            for (auto& matchPattern : matchPatterns)
+                [matchPatternsToRequest addObject:matchPattern->wrapper()];
+
+            [delegate webExtensionController:extensionController()->wrapper() promptForPermissionMatchPatterns:matchPatternsToRequest inTab:nil forExtensionContext:wrapper() completionHandler:originsCompletionHandler];
+        } else
+            completionHandler(false);
+    };
+
+    if ([delegate respondsToSelector:@selector(webExtensionController:promptForPermissions:inTab:forExtensionContext:completionHandler:)]) {
+        NSMutableSet *permissionsToRequest = [NSMutableSet set];
+        for (auto& permission : permissions)
```

Comment:
I think I might add a helper method in Cocoa helpers that translates a NSSet to HashSet and vice versa